src/structures/paging/page.rs

```diff
@@ -251,6 +252,43 @@ impl<S: PageSize> AddAssign<u64> for Page<S> {
     }
 }
 
+impl<S: PageSize> CheckedAdd<u64> for Page<S> {
+    type Output = Self;
+
+    /// Adds a number of pages to the left-hand side, checking for overflow.
+    /// A new `Page` is returned represented by the `Page` with a start address
+    /// equal to the sum of the left-hand side page's start address and the
+    /// `rhs` * `PageSize`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use x86_64::ops::CheckedAdd;
+    /// use x86_64::addr::VirtAddr;
+    /// use x86_64::structures::paging::{
+    ///     page::{Page, PageSize, Size4KiB}
+    /// };
+    ///
+    /// let start_addr_a = VirtAddr::zero();
+    /// let page_a: Page = Page::containing_address(start_addr_a);
+    /// let start_addr_b = VirtAddr::new(0x0 + Size4KiB::SIZE);
+    /// let page_b: Page = Page::containing_address(start_addr_b);
+    ///
+    /// assert_eq!(Some(page_b), page_a.checked_add(1u64));
+    /// assert_eq!(None, page_a.checked_add(u64::MAX));
+    /// ```
+    #[inline]
+    fn checked_add(self, rhs: u64) -> Option<Self::Output> {
+        let virt_addr_rhs = rhs
+            .checked_mul(S::SIZE)
+            .and_then(|addr| VirtAddr::try_new(addr).ok())?;
+
+        self.start_address()
+            .checked_add(virt_addr_rhs.as_u64())
```

Comment:
Why is `rhs` converted to a virtual address here? It's an offset, not an address. This is a problem because `rhs * S::SIZE` might be a valid offset, but not a valid address.